### DISCUSSION TITLE
Adding a flag for always pull/sync a submodule from the head revision of the selected branch

### DISCUSSION THREAD

**Discussion by eng-marani** [gitextensions/gitextensions#11587] [Ideas]
I know this is not a common workflow, but my goal is to always keep some specific submodule in sync automatically merging new contributions (from the head revision of a specific branch) instead of being linked to a specific submodule commit.
When the submodule has updates, i can see that looking on the tree where the "(nobranch)" label appear on submodule nodes.
Then, i always had to go into submodule and select/pull the correct branch, leading to mistakes.
Is it possible to automate this ? A new flag like "always sync to head revision" would be super useful for me in this scenario.

**Comment by vbjay** [comment]
Add a script that does this. You can add the correct script and run it on your own.

**Reply by vbjay** [reply]
Folowing problems:
- headless state is the norm for a submodule since submodule get set to right commit by checking out the sha in update init call. 
- Which branch did you want to sync off of?  What if not configured?

You will find that there are complexities. Why a script that does 
```shell
git submodule foreach --reccursive git checkout some branch

git submodule foreach --reccursive git pull
...
```
And potentially lets you pick remote and branch options is better. We do have script parameters available for that.

**Reply by vbjay** [reply]
See https://github.com/gitextensions/gitextensions/discussions/11509

**Comment by eng-marani** [comment]
"please do not do it" is not an option nor a valid reply for a requested workflow i think.
I know my needs have pitfalls, but i am not willing to execute that script manually each time.
Even more, your script wont properly work, because i have several submodules and only "some" of them i want to automatically always stay in the head revision, also each of these submods are checked out on a different branch name (not master).
At first, i saw that some of them were marked as "(no branch)", then i realized i had to go into submodule and pull, but without information about which was the previous right branch it was tied on. I know you can inspect the binded commit and understand which branch was previously in use, but this always make me think that this (the missing branch label on submodules) was a bug, even if it is not.
This is why there should be an advanced option, individually for each submodules, off by default, that allow to remember the branch i want to stay and automatically pull at the head commit of *that* branch, whenever the origin updates, pulling the parent.
Why should be so dangerous ?

**Reply by RussKie** [reply]
This scenario sounds very much custom to your workflow. A script which is triggered on "pull" is likely is better solution to your problem then trying to bake feature into the app which is unlikely to be used by the users.

**Reply by vbjay** [reply]
My script was an example. I did say it did more and that it could be improved with inputs and such. It was only posted as an example to let people see the direction and how to run multiple commands in a script. It was not a 100% solve the problem. It was just an example. Take it and fit to your needs. 

Yes the headless and which branch problems were kinda my point and exactly what I meant by being complex. You could try the git submodule update --remote option but again you would need to modify the submodule and set the branch to use.
